// node-googlehome/lib/connect.js
const googletts = require('google-tts-api'),
  Client = require('castv2-client').Client,
  DefaultMediaReceiver = require('castv2-client').DefaultMediaReceiver,
  mime = require('mime-types');

var ip;
var playerX;
function Connecter(ip) { this.ip = ip; this.lang = 'fr'
      this.player};

Connecter.prototype.readySpeaker = function(callback) {
      var client = new Client();
      client.connect(this.ip, function(){
        client.launch(DefaultMediaReceiver, function(err, player){
playerX =player;
if(err){

callback(err,player);
}else{

callback(0,player);
}
});
      });
    }
Connecter.prototype.config = function(options = {}) {
      this.lang = options.lang || this.lang
    }
Connecter.prototype.playMethod =function(url,callback) {

      if(!playerX) {
        this.readySpeaker()
console.log('=');
      }

      const params = {
        contentId: url,
        contentType: mime.lookup(url) || 'audio/mp3',
        streamType: 'BUFFERED'
      }

      return playerX.load(params, { autoplay: true }, function(err, status){
          if(err){
callback(err, status);
}else{
callback(0, status);
}
        });

    }
Connecter.prototype.playMedia=function(url,callback) {
      if(!url.startsWith('http')) throw new Error('This format is not supported.')
      this.playMethod(url,function(err,res){
callback(err,res);
});
    }
Connecter.prototype.speak=function(message, callback, speed = 1, timeout = 3000) {
     
      this.playMethod('http://translate.google.com/translate_tts?tl='+ this.lang+'&client=tw-ob&q='+encodeURI(message)+'&ie=UTF-8',function(err,res){
callback(err,res);
});
    }
module.exports=Connecter;
