// pratos_things_finder_plugin/index.js
const path = __dirname.replace('/node_modules/pratos_things_finder_plugin', '');
var user = require(path+'/lib/users.js');
var fs = require('fs');
 var style = require("pratos_style_class");
module.exports = {
    config: function () {

        var conf = Object();
        conf.GET = true;
       
        conf.navbar = true;
        conf.on_content = true;
        return conf;
    },
    navbar: function (nav) {
        nav.create_default_link("Things Finder", '<i class="fa fa-search" aria-hidden="true"></i>', '/pratos_things_finder/');
    },
    get: function (app, globalVariable) {

        app.get('/pratos_things_finder/', function (req, res) {
            res.setHeader('Content-Type', 'text/html');
          user.is_connected(req.user_id, function (response,user_res) {
			if (response == true) {
                    res.setHeader('Content-Type', 'text/html');
					   style.getColors(user_res, function (colors) {
                    require(path + "/node_modules/pratos_navbar_class").construct(globalVariable.navbarPlugins, function (res_nav) {
                        require(path+ "/node_modules/pratos_homepage_class").construct(globalVariable.homepagePlugins, function (res_homepage) {
                            res.render(__dirname + '/public/index.ejs', {
                                nav: res_nav,
                                content: globalVariable.contentPlugins,
                                homepage: res_homepage,
                                page_name: "pratos_things_finder",
								path: path,
								 global_color: colors.main,
                       
                        background_color: colors.second
							
                            });
                        });
                    });
					   });
                } else {
                    res.redirect('/');
                    delete globalVariable[req.user_id];
                }
            });
        });
        app.get('/things_finder.js', function (req, res) {
           fs.readFile(__dirname + '/public/things_finder.js', 'utf8', function(err, data) {
			   	res.setHeader('Content-Type', 'application/javascript');
			   res.end(data);
		   });
            
        });
    },
    content: function (link) {

        link["things_finder1"] = {
            "page": "admin/index",
            "script": " <script src='/things_finder.js'></script>"
        };

        return link;
    }
};
