// pratos_things_finder_plugin/index.js
const path = __dirname.replace('/node_modules/pratos_things_finder_plugin', '');
var User = require("pratos_user_class");
var fs = require('fs');

module.exports = {
    config: function () {

        var conf = Object();
        conf.GET = true;
       
        conf.navbar = true;
        conf.on_content = true;
        return conf;
    },
    navbar: function (nav) {
        nav.create_default_link("Things Finder", '<i class="fa fa-search" aria-hidden="true"></i>', '/pratos_things_finder/');
    },
    get: function (app, globalVariable) {

        app.get('/pratos_things_finder/', function (req, res) {
            res.setHeader('Content-Type', 'text/html');
            User.verify_connection(req.user_id, globalVariable, function (user_res) {
                if (user_res == true) {
                    res.setHeader('Content-Type', 'text/html');
					var global_color = 'rgb(' + globalVariable.colorStyle.r + "," + globalVariable.colorStyle.g + "," + globalVariable.colorStyle.b + ")";
		var background_color = 'rgb(' + globalVariable.secondColorStyle.r + "," + globalVariable.secondColorStyle.g + "," + globalVariable.secondColorStyle.b + ")";
                    require(path + "/node_modules/pratos_navbar_class").construct(globalVariable.navbarPlugins, function (res_nav) {
                        require(path+ "/node_modules/pratos_homepage_class").construct(globalVariable.homepagePlugins, function (res_homepage) {
                            res.render(__dirname + '/public/index.ejs', {
                                nav: res_nav,
                                content: globalVariable.contentPlugins,
                                homepage: res_homepage,
                                page_name: "pratos_things_finder",
								path: path,
								global_color: global_color,
			
			background_color: background_color
                            });
                        });
                    });

                } else {
                    res.redirect('/');
                    delete globalVariable[req.user_id];
                }
            });
        });
        app.get('/things_finder.js', function (req, res) {
           fs.readFile(__dirname + '/public/things_finder.js', 'utf8', function(err, data) {
			   	res.setHeader('Content-Type', 'application/javascript');
			   res.end(data);
		   });
            
        });
    },
    content: function (link) {

        link["things_finder1"] = {
            "page": "admin/index",
            "script": " <script src='/things_finder.js'></script>"
        };

        return link;
    }
};
