// pratos_style_class/index.js
/*
 * License MIT
 * « Copyright © 2018, Pratos »
 * v 1.5.0
 */
const path = __dirname.replace('/node_modules/pratos_style_class', '');
var user = require(path + '/lib/users.js');
module.exports = {
    updateColor: function (colors, client, callback) {
        global.globalVariable.database.mongoDB.collection("users").update({
            userID: client.userID
        }, {
            $set: {
                colors: {
                    main: colors.mainColor,
                    second: colors.secondColor
                }
            }
        }, function () {
            callback();
        });
    },
    getColors: function (client, callback) {
        global.globalVariable.database.mongoDB.collection("users").find({
            userID: client.userID
        }).toArray(function (err, result) {
            if (err)
                throw err;
            if (result[0].colors) {
                callback(result[0].colors)
            } else {
                callback({
                    main: "rgb(120, 126, 126)",
                    second: "rgb(126, 12, 12)"
                })
            }
        });
    }
};
